octree_web_viewer/src/bin/points_web_viewer.rs

```diff
@@ -12,44 +12,100 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use octree_web_viewer::utils;
-use point_viewer::octree;
+use octree_web_viewer::backend_error::PointsViewerError;
+use octree_web_viewer::state::AppState;
+use octree_web_viewer::utils::start_octree_server;
 use std::path::PathBuf;
 use std::sync::Arc;
 use structopt::StructOpt;
 
 /// HTTP web viewer for 3d points stored in OnDiskOctrees
 #[derive(StructOpt, Debug)]
 #[structopt(name = "points_web_viewer", about = "Visualizing points")]
-struct CommandlineArguments {
+pub struct CommandLineArguments {
     /// The octree directory to serve, including a trailing slash.
-    octree_path: String,
+    /// this overrides the path_* options
+    #[structopt(name = "DIR", parse(from_os_str))]
+    octree_path: Option<PathBuf>,
     /// Port to listen on.
-    #[structopt(default_value = "5433", long = "--port")]
+    #[structopt(default_value = "5433", long = "port")]
     port: u16,
     /// IP string.
-    #[structopt(default_value = "127.0.0.1", long = "--ip")]
+    #[structopt(default_value = "127.0.0.1", long = "ip")]
     ip: String,
+    /// instead of DIR: specify path prefix for octree dir
+    #[structopt(long = "prefix", parse(from_os_str), conflicts_with = "octree_path")]
+    path_prefix: Option<PathBuf>,
+    /// Optional: suffix for subfolder of octree dir
+    #[structopt(long = "suffix", parse(from_os_str))]
+    path_suffix: Option<PathBuf>,
+    /// instead of DIR: specify path folder for octree dir
+    #[structopt(long = "octree_id", conflicts_with = "octree_path")]
+    octree_id: Option<String>,
+    /// Cache items
+    #[structopt(default_value = "20", long = "cache_items")]
+    cache_max: usize,
+}
+
+/// init app state with command arguments
+/// backward compatibilty is ensured
+pub fn state_from(args: CommandLineArguments) -> Result<AppState, PointsViewerError> {
+    let mut has_suffix = false;
```

Comment:
How about removing `has_suffix` and later just checking for `!suffix.to_string_lossy().is_empty()`?

Reply:
done